substrate/frame/sassafras/src/tests.rs

```diff
@@ -0,0 +1,663 @@
+		assert_eq!(meta, TicketsMetadata::default());
+
+		tickets.iter().for_each(|(id, _)| {
+			let data = TicketsData::<Test>::get(id);
+			assert!(data.is_none());
+		});
+		// We used the last known next epoch randomness as a fallback
+		assert_eq!(next_random, Sassafras::randomness());
+	});
+}
+
+#[test]
+fn obsolete_tickets_are_removed_on_epoch_change() {
+	let (pairs, mut ext) = new_test_ext_with_pairs(4, false);
+	let pair = &pairs[0];
+	let start_slot = Slot::from(100);
+	let start_block = 1;
+
+	ext.execute_with(|| {
+		let epoch_duration: u64 = <Test as Config>::EpochDuration::get();
+
+		initialize_block(start_block, start_slot, Default::default(), pair);
+
+		let tickets = make_ticket_bodies(10, pair);
+		let mut epoch1_tickets = tickets[..4].to_vec();
+		let mut epoch2_tickets = tickets[4..].to_vec();
+
+		// Persist some tickets for next epoch (N)
+		persist_next_epoch_tickets(&epoch1_tickets);
+		assert_eq!(TicketsMeta::<Test>::get().tickets_count, [0, 4]);
+		// Check next epoch tickets presence
+		epoch1_tickets.sort_by_key(|t| t.0);
+		(0..epoch1_tickets.len()).into_iter().for_each(|i| {
+			let id = TicketsIds::<Test>::get((1, i as u32)).unwrap();
+			let body = TicketsData::<Test>::get(id).unwrap();
+			assert_eq!((id, body), epoch1_tickets[i]);
+		});
+
+		// Advance one epoch to enact the tickets
+		go_to_block(start_block + epoch_duration, start_slot + epoch_duration, pair);
+		assert_eq!(TicketsMeta::<Test>::get().tickets_count, [0, 4]);
+
+		// Persist some tickets for next epoch (N+1)
+		persist_next_epoch_tickets(&epoch2_tickets);
+		assert_eq!(TicketsMeta::<Test>::get().tickets_count, [6, 4]);
+		epoch2_tickets.sort_by_key(|t| t.0);
+		// Check for this epoch and next epoch tickets presence
+		(0..epoch1_tickets.len()).into_iter().for_each(|i| {
+			let id = TicketsIds::<Test>::get((1, i as u32)).unwrap();
+			let body = TicketsData::<Test>::get(id).unwrap();
+			assert_eq!((id, body), epoch1_tickets[i]);
+		});
+		(0..epoch2_tickets.len()).into_iter().for_each(|i| {
+			let id = TicketsIds::<Test>::get((0, i as u32)).unwrap();
+			let body = TicketsData::<Test>::get(id).unwrap();
+			assert_eq!((id, body), epoch2_tickets[i]);
+		});
+
+		// Advance to epoch 2 and check for cleanup
+
+		go_to_block(start_block + 2 * epoch_duration, start_slot + 2 * epoch_duration, pair);
+		assert_eq!(TicketsMeta::<Test>::get().tickets_count, [6, 0]);
+
+		(0..epoch1_tickets.len()).into_iter().for_each(|i| {
+			let id = TicketsIds::<Test>::get((1, i as u32)).unwrap();
+			assert!(TicketsData::<Test>::get(id).is_none());
+		});
+		(0..epoch2_tickets.len()).into_iter().for_each(|i| {
+			let id = TicketsIds::<Test>::get((0, i as u32)).unwrap();
+			let body = TicketsData::<Test>::get(id).unwrap();
+			assert_eq!((id, body), epoch2_tickets[i]);
+		});
+	})
+}
+
+// TODO davxy: create a read_tickets method which reads pre-constructed good tickets
+// from a file. Creating this stuff "on-the-fly" is just too much expensive
```

Comment:
Too expensive.

May be useful to add benchmarks for the ticket creation on the client side.